src/components/ExcelReader/ExcelReader.tsx

```diff
@@ -1,44 +1,90 @@
-import { useState } from "react";
+import { LinearProgress } from "@mui/material";
+import { ChangeEvent, useState } from "react";
+import { useRecoilCallback } from 'recoil';
+import excelCommunicator from "../../communication/excelCommunicator";
+import { notifySuccess } from "../../helpers/toaster";
 import { onLoad } from "../../helpers/xls-reader";
+import { ReactComponent as AddIcon } from "../../icons/addFile.svg";
+import { ReactComponent as ReplaceIcon } from "../../icons/replace-icon.svg";
+import { ReactComponent as UploadIcon } from "../../icons/upload.svg";
+import { dataSourcesAtom } from "../../recoil/dataSources/dataSources";
 import { Table } from "../../types/data";
+import { setState } from "../../types/utility";
 import './ExcelReader.scss';
 
 interface ExcelReaderProps {
-    addDataInstanceTable: (dataSourceId: string, table: Table, info: any) => void;
+    dashboardId: string;
+    isAdding: boolean;
+    dataSourceIdToReplace?: string;
+    setIsAdding: setState<boolean>;
 }
 
-const ExcelReader = ({ addDataInstanceTable }: ExcelReaderProps) => {
-    const [file, setFile] = useState<Blob>(null);
-    console.log(file);
+const ExcelReader = ({ dashboardId, isAdding, dataSourceIdToReplace, setIsAdding }: ExcelReaderProps) => {
+    const [file, setFile] = useState<File>(null);
+    const [isLoading, setIsLoading] = useState(false);
 
-    const filePathset = (e: any) => {
+    const filePathset = (e: ChangeEvent<HTMLInputElement>) => {
         e.stopPropagation();
         e.preventDefault();
         const file = e.target.files[0];
         setFile(file);
     };
 
     const readFile = () => {
+        setIsLoading(true);
         const reader = new FileReader();
-        reader.onload = (evt) => onLoad(evt, file, addDataInstanceTable);
+        reader.onload = (evt) => onLoad(dashboardId, evt, file, onReadingSucceed, onReadingEnd);
         reader.readAsBinaryString(file);
     };
 
+    const onReadingEnd = () => {
+        setFile(null);
+        setIsLoading(false);
+    };
+
+    const onReadingSucceed = useRecoilCallback(({ set }) => async (table: Table, fileName: string) => {
+        if (isAdding) {
```

Comment:
isAdding(dataSource)? like is this what it means?

Reply:
changed to fileUploadStage enum to make it clear